File: src/java/data_structures/Vector.java
package src.java.data_structures;
import java.lang.reflect.Array;


public class Vector<T> implements DataStructure<T>
{
	public Vector(final Class<T> clazz, int size)
	{
		mClazz = clazz;

		if (size <= 0)
			size = 10;
		
		reserve(size);
	}


	public int size()
	{
		return mSize;
	}


	public void reserve(final Integer need)
	{
		@SuppressWarnings("unchecked")
		T[] tmp = (T[]) Array.newInstance(mClazz, need);

		for (int i = 0; i < size(); ++i)
			tmp[i] = mData[i];

		mData = tmp;
	}


	public void pushBack(T value)
	{
		if (size() >= mData.length)
			reserve(size() + (size() / 2));

		mData[size()] = value;
		mSize += 1;
	}


	public T popBack()
	{
		mSize -= 1;
		T elem = mData[size()];
		return elem;
	}


	public T get(final int index)
	{
		return mData[index];
	}


	public void set(final int index, T value)
	{
		mData[index] = value;
	}


	private T[] mData = null;
	private Class<T> mClazz = null;
	private int mSize = 0;

}

File: src/csharp/DataStructures/Vector.cs


namespace DataStructures
{

	public class Vector<T> : IDataStructure<T>
	{
		private const int kInitialCap = 10;


		private int Cap { get; set; } = kInitialCap;
		private T[] _data = new T[kInitialCap];


		public int Count { get; internal set; } = 0;


		public T this[int index]
		{
			get
			{
				return Get(index);
			}
			set
			{
				Set(index, value);
			}
		}


		public T Get(int index)
		{
			return _data[index];
		}


		public void Set(int index, T elem)
		{
			_data[index] = elem;
		}


		public void PushBack(T elem)
		{
			if (Count >= Cap)
				Reserve(Count + (Count / 2));

			_data[Count] = elem;
			Count = Count + 1;
		}


		public T PopBack()
		{
			Count = Count - 1;
			T r = _data[Count];
			return r;
		}


		public void Reserve(int need)
		{
			if (need <= Cap)
				return;

			T[] tmp = new T[need];

			for (int i = 0; i < Count; ++i)
				tmp[i] = _data[i];

			_data = tmp;
			Cap = need;
		}

	}


}



